libexec/revng/revng-check-conventions

```diff
@@ -1,498 +1,727 @@
-    cat > /dev/stderr << END_USAGE_TEXT
-SYNOPSIS
-   $0 [<options>] [FILES...]
+executor = ThreadPoolExecutor()
+term_colors = {"reset": "\033[0;0m", "red": "\033[1;31m", "blue": "\033[1;34m"}
 
 
-DESCRIPTION
-   Checks that the files specified in FILES... respect the rev.ng coding
-   conventions, and prints all the violations.
+@dataclass
+class ProcessResult:
+    output: str
+    returncode: int
 
-   If the FILES... arguments are missing, checks that all the C++ source files in
-   the project that have already been added to the git repository project
-   respect the rev.ng coding conventions.
 
+@dataclass
+class PassResult:
+    output: str
+    ok: bool
 
-OPTIONS
-   --help
-       Prints this help and exit
+    def verify(self):
+        return self.output.strip() == "" and self.ok
 
-   --force-format
-       In addition to checking the rev.ng coding conventions on the specified
-       files, uses clang-format to try to enforce the coding conventions
-       automatically.
-       If this fails, exit with failure, otherwise exit with success.
-       WARNING: using this option will overwrite your files, make sure to backup
-       important stuff.
+    @staticmethod
+    def _check_processes(*args: ProcessResult) -> bool:
+        combined_stdout = "\n".join(p.output for p in args)
+        return all(p.returncode == 0 for p in args) and combined_stdout.strip() == ""
 
-    --use-local-clang-format-file
-      By default, clang-format is executed with a hard-coded configuration that
-      should apply to all the projects under the rev.ng umbrella.
-      When this option is passed, the hard-coded configuration is ignored, and
-      clang-format looks for configuration in a .clang-format file in the
-      current directory or in its parent directories.
+    @classmethod
+    def from_process(cls, *processes: ProcessResult) -> "PassResult":
+        return PassResult("\n".join(p.output for p in processes), cls._check_processes(*processes))
 
-      This allows to use different clang-format configuration on per-project
-      basis, in cases where the hard-coded configuration does not fit (e.g Qt
-      projects where Qt coding conventions are a better fit).
+    @staticmethod
+    def combine(*pass_results: "PassResult") -> "PassResult":
+        return PassResult(
+            "\n".join(p.output for p in pass_results), all(p.ok for p in pass_results)
+        )
 
-      See clang-format documentation for more details.
 
-    --print-clang-format-config
-      Print the clang-format configuration to stdout, ignoring all the other
-      arguments except for --use-local-clang-format-file.
-      The conventions are not checked.
+class Pass(abc.ABC):
+    def __init__(self, name: str, tag: str):
+        self.name = name
+        self.tag = tag
 
-    --commit-range <RANGE>
-      Check files changed in the specified commit range
+    def __call__(self, files: Dict[str, Dict[Path, str]], args: Dict[str, str]) -> PassResult:
```

Comment:
Overriding `__call__` usually makes things a bit obscure.
Is there any real benefit?
What's the difference between `__call__` and `process`?

Reply:
1. I thought it was fitting because you "run" the pass given a filtered file list and a list of arguments, if you prefer it can be renamed to `run`
2.I've added a bit of doc but basically ` __call__` will be given all the files and in its default implementation will check if there are any files for the `tag` attribute and if so will forward to the process function, this is so that a `Pass` can operate on more than one tag but in most cases it will be inherited and `process` will be overridden

Reply:
So `process` works on an individual `tag` while `__call__` works on a `Dict[str, List[str]]`?
Maybe a better name is `process_tag`.
As for `__call__`, operators override is scary and suspicious, I'd do it only if you need to interact with some library that expects an object to actually be callable. Otherwise a `.process` or `.run` method is simpler and more explicit.

Reply:
Renamed to `run`, kept `process`, `process_tag` sounds like you are doing an operation on the tags, not that you'll be given a filelist of a single tag